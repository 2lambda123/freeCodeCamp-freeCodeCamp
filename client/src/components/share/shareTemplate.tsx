import React from 'react';
import { faTwitter } from '@fortawesome/free-brands-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

export interface ShareTemplateProps {
  handleRedirectToTwitter: () => void;
}

export const ShareTemplate: React.FC<ShareTemplateProps> = ({
  handleRedirectToTwitter
}) => {
  return (
    <button data-testid="ShareTemplateWrapperTestID" className='btn fade-in' onClick={handleRedirectToTwitter}>
      <FontAwesomeIcon title="twitterIcon" icon={faTwitter} size='1x' />
    </button>
  );
};
