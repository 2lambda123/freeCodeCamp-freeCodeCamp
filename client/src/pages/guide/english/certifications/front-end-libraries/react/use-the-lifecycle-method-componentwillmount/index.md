---
title: Use the Lifecycle Method componentWillMount
---
## Use the Lifecycle Method componentWillMount

The challenge is a basic introduction to React's LifeCycle components. Thhis challenge is set with the motive of introducing the ``` componentWillMount ``` Life Cycle method of React's.

These methods are also called Lifecycle Hooks.

### Hint 1

Log something to the console in the ```react componentWillMount ``` method.

### Hint 2

The `` console.log(); `` statement is used to log to the browser console.

### Solution

Since you can log anything, use the ```javascript console.log(); ``` and pass in a string. The ``` componentWillMount ``` methos should look like:

```javascript
componentWillMount() {
  console.log("Component being mounted");
}
```
