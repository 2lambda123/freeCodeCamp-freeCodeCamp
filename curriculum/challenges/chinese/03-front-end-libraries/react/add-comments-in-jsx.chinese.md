---
id: 5a24bbe0dba28a8d3cbd4c5e
title: Add Comments in JSX
challengeType: 6
isRequired: false
videoUrl: ''
localeTitle: Add Comments in JSX
---

## Description
<section id='description'>
JSX 是一种可以编译成有效 JavaScript 的语法。有时，为了便于阅读，你可能需要在代码中添加注释。像大多数编程语言一样，JSX 也有自己的方法来实现这一点。
要将注释放在 JSX 中，可以使用<code>{/* */}</code>语法来包裹注释文本。
</section>

## Instructions
<section id='instructions'>
代码编辑器中的 JSX 元素与你在上一个挑战中创建的元素类似。在提供的<code>div</code>元素中的某处添加注释，而不修改现有的<code>h1</code>或<code>p</code>元素。
</section>

## Tests
<section id='tests'>

```yml
tests:
  - text: 常量<code>JSX</code>应该返回一个<code>div</code>元素。
    testString: assert(JSX.type === 'div', '常量<code>JSX</code>应该返回一个<code>div</code>元素。');
  - text: <code>div</code>应该包含一个<code>h1</code>标签作为第一个元素。
    testString: assert(JSX.props.children[0].type === 'h1', '<code>div</code>应该包含一个<code>h1</code>标签作为第一个元素。');
  - text: <code>div</code>应该包含一个<code>p</code>标签作为第二个元素。
    testString: assert(JSX.props.children[1].type === 'p', '<code>div</code>应该包含一个<code>p</code>标签作为第二个元素。');
  - text: <code>JSX</code>应该包含一个注释。
    testString: 'getUserInput => assert(getUserInput("index").includes("/*") && getUserInput("index").includes("*/"), "<code>JSX</code>应该包含一个注释。");'

```

</section>

## Challenge Seed
<section id='challengeSeed'>






    <div id='jsx-seed'>
    
```jsx
const JSX = (
  <div>
    <h1>This is a block of JSX</h1>
    <p>Here's a subtitle</p>
  </div>
);
    
```
</div>


### After Test
<div id='jsx-teardown'>

```jsx
ReactDOM.render(JSX, document.getElementById('root'))

```

</div>



</section>

## Solution
<section id='solution'>

```js
const JSX = (
<div>
  <h1>This is a block of JSX</h1>
  { /* this is a JSX comment */ }
  <p>Here's a subtitle</p>
</div>);
```

</section>
              