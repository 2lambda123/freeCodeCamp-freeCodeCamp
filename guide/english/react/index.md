---
title: React
---

# React

React is a JavaScript library. Applications built with React usually run on the client-side in a web browser as a Single Page Application. Applications of this kind only communicate with the server when necessary, which makes them very fast compared to traditional websites that force the user to wait for the server to render an entire page and send it back to the browser for changes.

React is used for building user interfaces - what the user sees on their screen and how they interact with your application. Instead of having one large template you break it up into smaller pieces known as components. This approach is called Modularity, which benefits the reusability and testability of your components.

- It's declarative: React uses a declarative paradigm that makes it more readable.
- It's efficient: React computes the minimal set of changes necessary to keep your DOM up-to-date.
- It's flexible: React allows the user to render one or many components to the browser.
- And it's compatible: React works well with many popular libraries and frameworks.

## Why learn React?

1. React involves Composition. Many components wrapping up the functionalities into an encapsulated container. Many popular websites use React as the View layer when implementing the MVC architectural pattern. Popular websites such as Facebook, Instagram, Khan Academy, Airbnb.

Example:
```jsx
const Component2 = () => {
  return <div />;
};
const Component3 = () => {
  return <div />;
};
const Component1 = () => {
  return (
    <div>
      <Component2 />
      <Component3 />
    </div>
  );
};

ReactDOM.render(<Component1 />, document.getElementById('app'));
```

2. React is Declarative. Declarative programming is a programming paradigm that expresses the logic of a computation without describing its control flow. Declarative programming comes with certain advantages such as reduced side effects (occurs when we modify any state or mutating something or making an API request), minimizing mutability (mostly abstracted), enhanced readability, and less bugs.

3. Unidirectional dataflow. User interface in React is controlled by the state, ss the state updates it also updates the user interface as well. So our user interface progresses as the state changes.

### Virtual DOM

React's magic comes from it's interpretation of the DOM and it's strategy for creating UIs.

React uses the virtual DOM to render an HTML tree virtually first, and then, every time a state changes and we get a new HTML tree that needs to be taken to the browser’s DOM, instead of writing the whole new tree React will only write the difference between the new tree and the previous tree (since React has both trees in memory). This process is known as Tree Reconciliation.

### Reconciliation

React has a smart diffing algorithm that it uses to only regenerate in its DOM node what actually needs to be regenerated while it keeps everything else as is. This diffing process is possible because of React’s virtual DOM.

Using the virtual DOM, React keeps the last DOM version in memory and when it has a new DOM version to take to the browser, that new DOM version will also be in memory, so React can compute the difference between the new and the old versions.

React will then instruct the browser to update only the computed diff and not the whole DOM node. No matter how many times we regenerate our interface, React will take to the browser only the new “partial” updates.

### More Information:

- [React Official Site](https://reactjs.org/)
- [React challenges on freeCodeCamp](https://learn.freecodecamp.org/front-end-libraries/react)
- [React Github](https://github.com/facebook/react)
- [Awesome React](https://github.com/enaqx/awesome-react)
