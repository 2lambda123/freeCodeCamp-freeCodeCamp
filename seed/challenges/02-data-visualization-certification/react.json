{
  "name": "React",
  "order": 14,
  "isComingSoon": "true",
  "time": "5h",
  "challenges": [
    {
      "id": "bd7158d8c423ede3aeb5bdee",
      "title": "Learn React Challenges",
      "description": [],
      "releasedOn": "October 1, 2014",
      "head": [],
      "challengeSeed": [],
      "tail": [],
      "solutions": [],
      "tests": [],
      "type": "waypoint",
      "MDNlinks": [],
      "challengeType": 3,
      "nameCn": "",
      "descriptionCn": [],
      "nameFr": "",
      "descriptionFr": [],
      "nameRu": "",
      "descriptionRu": [],
      "nameEs": "Crea una aplicaci√≥n de votaciones",
      "descriptionEs": [],
      "namePt": "",
      "descriptionPt": []
    }
  ]
}
