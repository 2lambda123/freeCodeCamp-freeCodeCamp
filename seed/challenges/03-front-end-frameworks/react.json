{
  "name": "React",
  "order": 5,
  "time": "5 hours",
  "helpRoom": "Help",
  "challenges": [
    {
      "id": "1d7123c8c441eddfaeb5bdef",
      "title": "",
      "description": [],
      "challengeSeed": [],
      "tests": [],
      "type": "waypoint",
      "challengeType": 0,
      "translations": {
        "de": {
          "description": [],
          "title": ""
        },
        "fr": {
          "description": [],
          "title": ""
        },
        "pt-br": {
          "description": [],
          "title": ""
        },
        "ru": {
          "description": [],
          "title": ""
        }
      }
    }
  ]
}