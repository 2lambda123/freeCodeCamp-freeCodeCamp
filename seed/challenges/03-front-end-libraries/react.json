{
  "name": "React",
  "order": 5,
  "time": "5 hours",
  "helpRoom": "Help",
  "challenges": [
    {
      "id": "587d7dbc367417b2b2512bb1",
      "title": "Introduction to the React Challenges",
      "description": [
        [
          "",
          "",
          "React is a JavaScript library designed to make it easier to build complex user interfaces. React does this in a few ways:<br><br><ol><li>React allows HTML to render from a JavaScript file.</li><li>React breaks a website's parts into Components that can then be combined to build pages.</li><li>React Components can pass data to each other as properties</li></ol><br><br>React uses a virtual Document Object Model. Instead of changing the DOM directly, React makes a simplified copy of the DOM (the virtual DOM) when the page loads. This allows React to make changes to elements on the page without re-rendering the entire page. This can improve the responsiveness and functionality of single-page applications.<br><br>While not part of the React library and not required, JSX is often paired with React. JSX is a preprocessor for JavaScript written in a syntax akin to XML that results in more readable code.<br><br><strong>Fun fact:</strong> freeCodeCamp is built using React.",
          ""
        ],
        [
          "",
          "",
          "The React challenges have not been ported into freeCodeCamp yet. You can visit <a href=\"http://hysterical-amusement.surge.sh/\" target=\"_blank\">this link to work through the alpha version</a> of these challenges. If you have feedback, you can open an issue (or pull request) directly on <a href=\"https://github.com/bonham000/fcc-react-tests-module\" target=\"_blank\">this repository.</a>",
          ""
        ]
      ],
      "releasedOn": "Feb 17, 2017",
      "challengeSeed": [],
      "tests": [],
      "type": "waypoint",
      "challengeType": 7,
      "isRequired": false,
      "translations": {}
    },
    {
      "id": "58b1157884fee96bd6b13851",
      "title": "Create a Simple JSX Element",
      "description": [
        "React is an Open Source view library created and maintained by Facebook.",
        "It's a great tool to render the User Interface (UI) of modern web applications.<br><br>",
        "React uses a syntax extension of JavaScript called JSX that allows you to write HTML directly within JavaScript.",
        "This has several benefits. It lets you use the full programmatic power of JavaScript within HTML, and helps to keep your code readable.",
        "For the most part, JSX is similar to the HTML that you have already learned, however there are a few key differences that will be covered throughout these challenges.<br><br>",
        "For instance, because JSX is a syntactic extension of JavaScript, you can actually write JavaScript directly within JSX.",
        "To do this, you simply include the code you want to be treated as JavaScript within curly braces: <code>{ 'this is treated as JavaScript code' }</code>.",
        "Keep this in mind, since it's used in several future challenges.<br><br>",
        "However, because JSX is not valid JavaScript, JSX code must be compiled into JavaScript.",
        "The transpiler Babel is a popular tool for this process. For your convenience, it's already added behind the scenes for these challenges.",
        "If you happen to write syntactically invalid JSX, you will see the first test in these challenges fail.",
        "The current code uses JSX to assign a <code>div</code> element to the constant <code>JSX</code>.",
        "Replace the <code>div</code> with an <code>h1</code> element",
        "and add the text <code>Hello JSX!</code> inside it."
      ],
      "files": {
        "indexjsx": {
          "content": "const JSX = <div></div>;",
          "head": "(function() {",
          "key": "indexjsx",
          "path": "index.jsx",
          "tail": ";\n return JSX })()"
        }
      },
      "tests": [
        {
          "test": "assert.strictEqual(jsx.type, 'h1', 'The constant JSX should return an h1 element.');",
          "text": "The constant JSX should return an h1 element."
        },
        {
          "test": "assert.strictEqual(jsx.props.children, 'Hello JSX!', true, 'The h1 tag should include the text 'Hello JSX!');",
          "text": "The h1 tag should include the text 'Hello JSX!'"
        }
      ],
      "type": "modern",
      "isRequired": true,
      "translations": {}
    }
  ]
}
