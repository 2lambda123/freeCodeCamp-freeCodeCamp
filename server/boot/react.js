import React from 'react';
import debug from 'debug';
import { renderToString } from 'react-dom/server';
import createMemoryHistory from 'history/lib/createMemoryHistory';

import { createApp, provideStore, App } from '../../common/app';
import waitForEpics from '../../common/utils/wait-for-epics.js';
import { titleSelector } from '../../common/app/redux';

const log = debug('fcc:react-server');

// add routes here as they slowly get reactified
// remove their individual controllers
const routes = [
  '/challenges',
  '/challenges/*',
  '/map',
  '/settings',
  '/settings/*'
];

const devRoutes = [];

export default function reactSubRouter(app) {
  var router = app.loopback.Router();

  router.get('/videos', (req, res) => res.redirect('/map'));
  router.get(
    '/videos/:dashedName',
    (req, res) => res.redirect(`/challenges/${req.params.dashedName}`)
  );

  // These routes are in production
  routes.forEach((route) => {
    router.get(route, serveReactApp);
  });

  if (process.env.NODE_ENV === 'development') {
    devRoutes.forEach(function(route) {
      router.get(route, serveReactApp);
    });
  }

  app.use('/:lang', router);

  function serveReactApp(req, res, next) {
    const { lang } = req;
    const serviceOptions = { req };
    createApp({
      serviceOptions,
      history: createMemoryHistory({ initialEntries: [ req.originalUrl ] }),
      defaultStaet: { app: { lang } }
    })
      .filter(({ redirect, notFound }) => {
        if (redirect) {
          log('found a redirect');
          res.redirect(redirect.pathname + redirect.search);
          return false;
        }

        if (notFound) {
          log(`tried to find ${req.path} but got 404`);
          next();
          return false;
        }

        return true;
      })
      .flatMap(({ store, epic }) => {
        return waitForEpics(epic)
          .map(() => renderToString(
            provideStore(React.createElement(App), store)
          ))
          .map((markup) => ({ markup, store, epic }));
      })
      .do(({ markup, store, epic }) => {
        log('react markup rendered, data fetched');
        const state = store.getState();
        const title = titleSelector(state);
        epic.dispose();
        res.expose(state, 'data');
        res.expose(req.flash(), 'flash');
        res.render('layout-react', { markup, title });
      })
      .subscribe(() => log('html rendered and sent'), next);
  }
}
