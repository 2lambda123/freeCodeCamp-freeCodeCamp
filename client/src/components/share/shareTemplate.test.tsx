import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { ShareTemplate } from './shareTemplate';

test('share template snapshot before copied', () => {
  render(<ShareTemplate handleClick={jest.fn()} />);
  const shareTemplate = screen.getByTestId('share-template');
  expect(shareTemplate).toMatchSnapshot();
});

test('testing share templete when isCopied', () => {
  const clickFn = jest.fn();
  render(<ShareTemplate handleClick={clickFn} />);

  const isCopyToShareButtonExists = screen.getByTestId(/copy-button/i);
  expect(isCopyToShareButtonExists).toHaveTextContent(
    /buttons.copy-link-to-share/i
  );
  fireEvent.click(isCopyToShareButtonExists);
  expect(clickFn).toBeCalledTimes(1);
});
