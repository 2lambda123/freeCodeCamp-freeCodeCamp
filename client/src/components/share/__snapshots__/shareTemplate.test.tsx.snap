// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`share template snapshot before copied 1`] = `
<button
  class="btn-block btn"
  data-testid="share-template"
>
  <span
    data-testid="copy-button"
  >
     
    buttons.copy-link-to-share
     
  </span>
</button>
`;
